Fix issue #4962, remove trailing whitespace in caption field when set to `uploadMediaDetails`

Jiaweeeeeen opened the pull request:
**Description (required)**

Fixes #4962

The caption field is used for the media filename and the MediaWiki has restrictions on the filename, thus the trailing whitespace causes the logcat error `titleblacklist-custom-space`. See [https://commons.wikimedia.org/wiki/MediaWiki:Titleblacklist](url).

This issue is fixed by checking and removing any trailing whitespace in the caption field. 
In `UploadMediaDetailAdapter.java`, any trailing space in`captionText` will be removed using `removeTrailingWhitespace` when set to `uploadMediaDetails`.

**Tests performed (required)**

Tested Build variant: BetaDebug on Pixel 2 with API level 30.

Need help? See https://support.google.com/android/answer/9075928

---

_Note: Please ensure that you have read CONTRIBUTING.md if this is your first pull request._

nicolas-raoul commented on app/src/test/kotlin/fr/free/nrw/commons/upload/UploadMediaDetailAdapterUnitTest.kt:
```diff
@@ -247,4 +247,16 @@ class UploadMediaDetailAdapterUnitTest {
 
+    @Test
+    fun testRemoveTrailingWhitespace() {
+        val test1 = "test  "
```
Thanks a lot for the unit test too! Would you mind testing a few more whitespace characters, for instance tab, carriage return, Japanese whitespace? And also without whitespace. Thanks!

Jiaweeeeeen commented:
Tests added.

nicolas-raoul reviewed (changes requested):
Hi!
I just tested this pull request.
Unfortunately, I still can't type Japanese spaces as a caption (I can type them in other apps).
Screencast: https://drive.google.com/file/d/1SuFzgBcAnjoppXz30e0PQ7ZK8_ub7lHk/view?usp=sharing
Would you mind checking?
Thanks!

Jiaweeeeeen commented:
I think that's because the issue #4985 haven't been resolved. Currently,  the Japanese space is still blocked by the input filter. I tested the Japanese trailing space in the unit tests and it can be removed. One of my group member had a solution to stop blocking the Japanese space and transfer it into English space, but he was not the assignee. I could ask him to open a PR and hope his solution could solve the problem.

Jiaweeeeeen commented:
Hi @nicolas-raoul, after talked to my teammate, I removed the blocking of Japanese Space with the pattern `\x{3000}` in `UploadMediaDetailInputFilter` and now you can type Japanese space in the caption field. I also added a method in `UploadMediaDetailAdapter` to convert the Japanese space into English Space before updating to `uploadMediaDetails`, so the upload will not fail. Could you test the pull request again?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:
```diff
@@ -418,10 +419,45 @@ private void setUpRecentLanguagesSection(final List<Language> recentLanguages) {
+         * @param source the source text
+         * @return a string with English space instead of Japanese space
+         */
+        public String convertJapSpaceToEngSpace(String source) {
```
Could you please rename from `convertJapSpaceToEngSpace` to `convertIdeographicSpaceToLatinSpace`? Thanks!

Jiaweeeeeen replied:
Done

nicolas-raoul reviewed (changes requested):
Just tested, it works great, thanks! https://commons.wikimedia.org/wiki/File:%E3%82%B5%E3%83%83%E3%83%9D%E3%83%AD_%E6%81%B5%E6%AF%94%E5%AF%BF%E3%82%AC%E3%83%BC%E3%83%87%E3%83%B3%E3%83%97%E3%83%AC%E3%82%A4%E3%82%B9.jpg
Please just rename the method and I will merge.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:
```diff
@@ -418,10 +419,45 @@ private void setUpRecentLanguagesSection(final List<Language> recentLanguages) {
+         * @return a string with Latin spaces instead of Ideographic spaces
+         */
+        public String convertIdeographicSpaceToLatinSpace(String source) {
+            Pattern JapSpacePattern = Pattern.compile("\\x{3000}");
```
Ah sorry I missed this one, could you please rename `JapSpacePattern` to `ideographicSpacePattern`?

Jiaweeeeeen replied:
Sorry about that, now changed.
